app/controllers/miq_policy_controller/policies.rb

```diff
@@ -1,91 +1,113 @@
 module MiqPolicyController::Policies
   extend ActiveSupport::Concern
 
-  def policy_edit
-    case params[:button]
-    when "cancel"
-      id = params[:id] ? params[:id] : "new"
-      return unless load_edit("policy_edit__#{id}", "replace_cell__explorer")
-      @policy = MiqPolicy.find_by(:id => @edit[:policy_id]) if @edit[:policy_id]
-      if @policy && @policy.id
-        add_flash(_("Edit of %{model} \"%{name}\" was cancelled by the user") % {:model => ui_lookup(:model => "MiqPolicy"), :name => @policy.description})
-      else
-        add_flash(_("Add of new %{models} was cancelled by the user") % {:models => ui_lookup(:model => "MiqPolicy")})
+  def policy_edit_cancel
+    id = params[:id] ? params[:id] : "new"
+    return unless load_edit("policy_edit__#{id}", "replace_cell__explorer")
+    @policy = MiqPolicy.find_by(:id => @edit[:policy_id]) if @edit[:policy_id]
+    if @policy && @policy.id
+      add_flash(_("Edit of %{model} \"%{name}\" was cancelled by the user") % {:model => ui_lookup(:model => "MiqPolicy"), :name => @policy.description})
+    else
+      add_flash(_("Add of new %{models} was cancelled by the user") % {:models => ui_lookup(:model => "MiqPolicy")})
+    end
+    @edit = nil
+    get_node_info(x_node)
+    replace_right_cell(:nodetype => @nodetype)
+  end
+
+  def policy_edit_reset
+    @sb[:action] = "policy_edit"
+    policy_build_edit_screen(session[:edit].try(:key?, :typ) ? session[:edit][:typ] : params[:typ])
+    if params[:button] == "reset"
+      add_flash(_("All changes have been reset"), :warning)
+    end
+    replace_right_cell(:nodetype => "p")
+  end
+
+  def policy_edit_save
+    assert_privileges("policy_#{@policy.id ? "edit" : "new"}")
+    policy = @policy.id.blank? ? MiqPolicy.new : MiqPolicy.find(@policy.id) # Get new or existing record
+    policy.mode = @edit[:new][:mode]
+    policy.updated_by = session[:userid]
+
+    if @policy.id.blank?
+      policy.towhat = @edit[:new][:towhat]
+      policy.created_by = session[:userid]
+    end
+
+    case @edit[:typ]
+    when "basic"
+      policy.description = @edit[:new][:description]
+      policy.active = @edit[:new][:active]
+      policy.notes = @edit[:new][:notes]
+      policy.expression = @edit[:new][:expression]["???"] ? nil : MiqExpression.new(@edit[:new][:expression])
+    when "conditions"
+      mems = @edit[:new][:conditions].invert                  # Get the ids from the member list box
+      policy.conditions.collect { |pc| pc }.each { |c| policy.conditions.delete(c) unless mems.keys.include?(c.id) }  # Remove any conditions no longer in members
+      mems.each_key { |m| policy.conditions.push(Condition.find(m)) unless policy.conditions.collect(&:id).include?(m) }    # Add any new conditions
+    end
+
+    if !policy.valid? || @flash_array || !policy.save
+      policy.errors.each do |field, msg|
+        add_flash("#{field.to_s.capitalize} #{msg}", :error)
```

Comment:
probably not related to this pr, but why are we using flash to show field errors?

Reply:
why? I have no idea. Probably that was the way to do it 8 years ago or so when it was written.